### PR TITLE
add support for STR bulk packing

### PR DESCRIPTION
**Pull request by ambling**
For Issue #45 

add support of Rtree creation through STR packing, according to the paper:
[STR: a simple and efficient algorithm for R-tree packing](http://ieeexplore.ieee.org/abstract/document/582015/)

add a recursive function in RTree.Builder and set the default filling factor as 0.7 (1.0 is better for readonly RTree while 0.7 reserves space for insertion).

note that this implementation is only for 2D RTree.

### CODE REVIEW

**Comment by ambling** [comment]
Opps...

I am sorry the CI checks failed..

However, I cannot find any clue from the fail log about why the tests crash on Java6

**Comment by davidmoten** [comment]
Thanks for the PR, looks interesting! I'll review this weekend.

**Comment by ambling** [comment]
I fixed a bug but the CI failed again with openjdk6.

After trying the updated workaround on travis-ci/travis-ci#5227 it seems to be working.

**Comment by davidmoten** [comment]
What's your performance conclusion about using bulk loading?

**Comment by ambling** [comment]
@davidmoten 
the results on index creation time in the benchmark on my laptop:
```
Benchmark                                                                               Mode  Cnt        Score       Error  Units
BenchmarksRTree.bulkLoadingFullRTreeCreation010                                        thrpt   10       73.747 ±     3.650  ops/s
BenchmarksRTree.bulkLoadingRTreeCreation010                                            thrpt   10       86.520 ±     1.905  ops/s
BenchmarksRTree.defaultRTreeCreation010                                                thrpt   10        6.506 ±     0.062  ops/s
BenchmarksRTree.flatBufferRTreeCreation010                                             thrpt   10     3818.918 ±    65.382  ops/s
BenchmarksRTree.starRTreeCreation010                                                   thrpt   10        3.141 ±     0.047  ops/s
```

The bulk loading is about 10x faster than the default RTree creation and about 20x faster than the R*-tree creation. 

The FlatBuffers-based RTree is much faster is because the creation process just load the basic context info and one root node.

Besides, it is interesting in the results that bulk packing with 70% full nodes is a little faster than that with 100% full nodes, but there would be less nodes created if the nodes are 100% full. I may dig this deeper for the reason.

As for the insertion/deletion/search performance on the STR-RTree, it would be reasonably worse than that of the R*-tree, but I haven't check this yet.

And for geometry types other than points, the performance of STR-RTree maybe even more worse, because the packing is based on just the central position, leading to more overlapping among nodes.

File: src/main/java/com/github/davidmoten/rtree/RTree.java [hunk 1]
```diff
@@ -292,16 +321,106 @@ public Builder factory(Factory<?, ? extends Geometry> factory) {
...
+            }
+
+            private float mid(HasGeometry o) {
+                Rectangle mbr = o.geometry().mbr();
+                if (dimension == 0) return (mbr.x1() + mbr.x2()) / 2;
+                else return (mbr.y1() + mbr.y2()) / 2;
+            }
+
+            @Override
+            public boolean equals(Object obj) {
```
Comment (davidmoten) [inline src/main/java/com/github/davidmoten/rtree/RTree.java]: why are you implementing `equals`? I'd say remove.

**Comment by ambling** [comment]
Thanks for the review!

I have made the changes, please have a look at that.

**Comment by davidmoten** [comment]
Thanks @ambling. So I suppose the remaining task is to come up with some reasonably comprehensive advice about the pros and cons of using the STR bulk load. Would you be able to do a bit more perf testing so that we have something concrete to advise users?

**Comment by ambling** [comment]
Yes @davidmoten . I think maybe I can add one or more test datasets and more benchmarking cases to verify the performance. I may finish the task in several days.

**Comment by davidmoten** [comment]
Thanks @ambling. Let's merge this now and when you have more detail we'll add it the docs. Thanks very much for your work!

**Pull request merged**